File: apps/minecraft/spigot/src/main/java/crabcraft/net/crabUtilities/heads/PlayerHeadDropsListener.java
package crabcraft.net.crabUtilities.heads;

import crabcraft.net.crabUtilities.CrabUtilities;
import crabcraft.net.crabUtilities.NicknameComponentResolver;
import net.kyori.adventure.text.Component;
import net.kyori.adventure.text.format.NamedTextColor;
import net.kyori.adventure.text.format.TextDecoration;
import org.bukkit.Material;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.EventPriority;
import org.bukkit.event.Listener;
import org.bukkit.event.entity.PlayerDeathEvent;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.meta.SkullMeta;

import java.time.LocalDate;
import java.time.format.TextStyle;
import java.util.List;
import java.util.Locale;

public class PlayerHeadDropsListener implements Listener {

    private final CrabUtilities plugin;

    public PlayerHeadDropsListener(final CrabUtilities plugin) {
        this.plugin = plugin;
    }

    @EventHandler(priority = EventPriority.HIGH)
    public void onPlayerDeath(final PlayerDeathEvent event) {
        if (!this.plugin.getConfig().getBoolean("tweaks.player-head-drops.enabled", false)) {
            return;
        }
        final Player victim = event.getEntity();
        final Player killer = victim.getKiller();
        if (killer == null) {
            return;
        }

        final ItemStack head = new ItemStack(Material.PLAYER_HEAD);
        if (!(head.getItemMeta() instanceof final SkullMeta meta)) {
            return;
        }
        meta.setOwningPlayer(victim);
        final Component victimName = NicknameComponentResolver.forPlayer(this.plugin.getEssentials(), victim);
        meta.displayName((victimName != null ? victimName : Component.text(victim.getName(), NamedTextColor.GOLD))
                .append(Component.text("'s Head", NamedTextColor.GOLD))
                .decoration(TextDecoration.ITALIC, false));
        final Component killerName = NicknameComponentResolver.forPlayer(this.plugin.getEssentials(), killer);
        meta.lore(List.of(Component.text("Killed by ", NamedTextColor.GRAY)
                .append(killerName != null ? killerName : Component.text(killer.getName(), NamedTextColor.GRAY))
                .decoration(TextDecoration.ITALIC, false),
                Component.text("Killed on: " + formatDate(LocalDate.now()), NamedTextColor.GRAY)
                        .decoration(TextDecoration.ITALIC, false)));
        head.setItemMeta(meta);
        event.getDrops().add(head);
    }

    private static String formatDate(final LocalDate date) {
        final int day = date.getDayOfMonth();
        final String suffix = switch (day) {
            case 1, 21, 31 -> "st";
            case 2, 22 -> "nd";
            case 3, 23 -> "rd";
            default -> "th";
        };
        return day + suffix + " " + date.getMonth().getDisplayName(TextStyle.FULL, Locale.ENGLISH);
    }
}

File: apps/minecraft/spigot/src/main/kotlin/crabcraft/net/crabUtilities/heads/PlayerHeadDropsListener.kt
package crabcraft.net.crabUtilities.heads

import crabcraft.net.crabUtilities.CrabUtilities
import crabcraft.net.crabUtilities.NicknameComponentResolver
import java.time.LocalDate
import java.time.format.TextStyle
import java.util.Locale
import net.kyori.adventure.text.Component
import net.kyori.adventure.text.format.NamedTextColor
import net.kyori.adventure.text.format.TextDecoration
import org.bukkit.Material
import org.bukkit.event.EventHandler
import org.bukkit.event.EventPriority
import org.bukkit.event.Listener
import org.bukkit.event.entity.PlayerDeathEvent
import org.bukkit.inventory.ItemStack
import org.bukkit.inventory.meta.SkullMeta

open class PlayerHeadDropsListener(private val plugin: CrabUtilities) : Listener {
    @EventHandler(priority = EventPriority.HIGH)
    open fun onPlayerDeath(event: PlayerDeathEvent) {
        if (!plugin.config.getBoolean("tweaks.player-head-drops.enabled", false)) return
        val victim = event.entity
        val killer = victim.killer ?: return
        val head = ItemStack(Material.PLAYER_HEAD)
        val meta = head.itemMeta as? SkullMeta ?: return
        meta.owningPlayer = victim
        val victimName = NicknameComponentResolver.forPlayer(plugin.getEssentials(), victim)
        meta.displayName(
            (victimName ?: Component.text(victim.name, NamedTextColor.GOLD))
                .append(Component.text("'s Head", NamedTextColor.GOLD))
                .decoration(TextDecoration.ITALIC, false)
        )
        val killerName = NicknameComponentResolver.forPlayer(plugin.getEssentials(), killer)
        meta.lore(
            listOf(
                Component.text("Killed by ", NamedTextColor.GRAY)
                    .append(killerName ?: Component.text(killer.name, NamedTextColor.GRAY))
                    .decoration(TextDecoration.ITALIC, false),
                Component.text("Killed on: ${formatDate(LocalDate.now())}", NamedTextColor.GRAY)
                    .decoration(TextDecoration.ITALIC, false),
            )
        )
        head.itemMeta = meta
        event.drops.add(head)
    }

    private fun formatDate(date: LocalDate): String {
        val day = date.dayOfMonth
        val suffix =
            when (day) {
                1,
                21,
                31 -> "st"
                2,
                22 -> "nd"
                3,
                23 -> "rd"
                else -> "th"
            }
        return "$day$suffix ${date.month.getDisplayName(TextStyle.FULL, Locale.ENGLISH)}"
    }
}
